### PR TITLE
stream: fix `select()` race condition

### PR DESCRIPTION
**Pull request by indutny**
Call `uv__read()`/`uv__write()` before re-entering `select` thread.
Otherwise we might get second event from it, and will block indefinitely
on `uv__read()`/`uv__write()`.

Fix: https://github.com/libuv/libuv/issues/304

cc @saghul

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Ah, a race condition.  LGTM.

File: src/unix/stream.c [hunk 1]
```diff
@@ -242,6 +241,14 @@ static void uv__stream_osx_select_cb(uv_async_t* handle) {
 
   if ((events & UV__POLLOUT) && uv__io_active(&stream->io_watcher, UV__POLLOUT))
     uv__stream_io(stream->loop, &stream->io_watcher, UV__POLLOUT);
+
+  if (stream->flags & UV_CLOSING)
```
Comment (bnoordhuis) [inline src/unix/stream.c]: Suggestion: maybe use `uv__is_closing(stream)` here?

**Comment by secrettriangle** [comment]
Is it too complicated to add a test case for this? It's a race condition so it might be hard to test for this behavior but maybe even a probabilistic check would suffice?

**Comment by rauchg** [comment]
Just tested it. Fixes longstanding issues with stdout buffering on [clif](https://github.com/rauchg/clif).

**Comment by indutny** [comment]
@secrettriangle yeah, I think it'll be quite complicated. The issue reproduces only on debug build of node.js on my machine. I wasn't able to reproduce it with just libuv.

But if anyone has a suggestion for the test - I'd enjoy reviewing a PR with it ;)

**Comment by indutny** [comment]
Landed in 7005d29, thank you!

**Comment by saghul** [comment]
Great job Fedor!

**Comment by indutny** [comment]
Thank you, @saghul

**Pull request closed without merging**